File: mcp/conflux_mcp_demo/contract.py
from web3 import Web3
import json
import statistics

class IdentityContract:
    def __init__(self, rpc_url, contract_address, abi_path="IdentityRegistry.json"):
        self.w3 = Web3(Web3.HTTPProvider(rpc_url))
        with open(abi_path) as f:
            artifact = json.load(f)
        abi = artifact["abi"] if isinstance(artifact, dict) and "abi" in artifact else artifact
        self.contract = self.w3.eth.contract(address=contract_address, abi=abi)

    def _raw_tx_bytes(self, signed):
        # eth-account changed attribute name across versions
        raw = getattr(signed, "rawTransaction", None)
        if raw is None:
            raw = getattr(signed, "raw_transaction", None)
        if raw is None:
            # Some versions expose .rawTransaction as bytes-like mapping
            try:
                raw = signed["rawTransaction"]  # type: ignore[index]
            except Exception:
                pass
        if raw is None:
            raise AttributeError("SignedTransaction missing raw transaction bytes")
        return raw

    def register(self, agent_id, metadata, private_key):
        if not private_key:
            raise ValueError("private_key is required to sign the transaction")
        # Use the provided key as the signer so msg.sender == agent address
        local_account = self.w3.eth.account.from_key(private_key)
        chain_id = self.w3.eth.chain_id
        latest_block = self.w3.eth.get_block("latest")
        eip1559_supported = isinstance(latest_block, dict) and ("baseFeePerGas" in latest_block)
        # Build transaction with EIP-1559 if supported, otherwise legacy gasPrice
        common = {
            "from": local_account.address,
            "nonce": self.w3.eth.get_transaction_count(local_account.address),
            "gas": 1000000,
            "chainId": chain_id,
        }
        if eip1559_supported:
            try:
                # Use eth_feeHistory to estimate median tip and a robust fee cap
                fh = self.w3.eth.fee_history(10, "latest", [50])
                rewards = [r[0] for r in fh.get("reward", []) if isinstance(r, (list, tuple)) and len(r) > 0]
                if rewards:
                    tip = int(statistics.median(rewards))
                else:
                    tip = Web3.to_wei(2, "gwei")
                base_fees = fh.get("baseFeePerGas", [])
                if base_fees:
                    # Next block suggested base fee is the last element
                    next_base = int(base_fees[-1])
                else:
                    next_base = int(latest_block["baseFeePerGas"])
                # Enforce a sane minimum tip
                if tip < Web3.to_wei(1, "gwei"):
                    tip = Web3.to_wei(1, "gwei")
                # Robust cap: 2x next base fee + median tip
                max_fee = next_base * 2 + tip
            except Exception:
                # Fallback to simple heuristic if feeHistory is unavailable
                base_fee = int(latest_block["baseFeePerGas"])
                tip = Web3.to_wei(2, "gwei")
                max_fee = base_fee * 2 + tip
            fee_fields = {
                "maxPriorityFeePerGas": tip,
                "maxFeePerGas": max_fee,
                "type": 2,
            }
            tx = self.contract.functions.register(agent_id, metadata).build_transaction({**common, **fee_fields})
        else:
            legacy_fields = {
                "gasPrice": self.w3.eth.gas_price,
            }
            tx = self.contract.functions.register(agent_id, metadata).build_transaction({**common, **legacy_fields})
        signed = local_account.sign_transaction(tx)
        raw = self._raw_tx_bytes(signed)
        tx_hash = self.w3.eth.send_raw_transaction(raw)
        return tx_hash.hex()

    def get_entry(self, address):
        return self.contract.functions.getEntry(address).call()

    def get_all_entries(self, from_block: int = 0, to_block: str | int = "latest"):
        # Scan Registered events to discover all owners, then read current entries
        topic0 = self.w3.keccak(text="Registered(address,string,string,uint256)").hex()
        address = Web3.to_checksum_address(self.contract.address)
        logs = self.w3.eth.get_logs({
            "address": address,
            "fromBlock": from_block,
            "toBlock": to_block,
            "topics": [topic0],
        })
        owners_lower = []
        seen = set()
        for log in logs:
            # owner is the first indexed topic (topics[1])
            t1 = log["topics"][1].hex()
            owner = Web3.to_checksum_address("0x" + t1[-40:])
            low = owner.lower()
            if low not in seen:
                seen.add(low)
                owners_lower.append(owner)
        entries = []
        for owner in owners_lower:
            try:
                e = self.contract.functions.getEntry(owner).call()
                entries.append(e)
            except Exception:
                continue
        return entries

File: mcp_xmtp/src/contract.ts
import { Web3 } from 'web3';
import * as fs from 'fs-extra';
import { ContractEntry } from './types';

export class IdentityContract {
  private web3: Web3;
  private contract: any;

  constructor(rpcUrl: string, contractAddressOrFile: string, abiPath: string = 'IdentityRegistry.json') {
    this.web3 = new Web3(rpcUrl);
    
    // Read contract address from file if it's a file path
    let contractAddress: string;
    if (contractAddressOrFile.includes('/') || contractAddressOrFile.includes('\\')) {
      // It's a file path
      contractAddress = fs.readFileSync(contractAddressOrFile, 'utf8').trim();
    } else {
      // It's a direct address
      contractAddress = contractAddressOrFile;
    }
    
    const artifact = fs.readJsonSync(abiPath);
    const abi = artifact.abi || artifact;
    
    this.contract = new this.web3.eth.Contract(abi, contractAddress);
  }

  async getBalance(address: string): Promise<bigint> {
    return await this.web3.eth.getBalance(address);
  }

  private getRawTxBytes(signed: any): string {
    // Handle different versions of eth-account
    return signed.rawTransaction || signed.raw_transaction || signed['rawTransaction'];
  }

  async register(agentId: string, metadata: string, privateKey: string): Promise<string> {
    if (!privateKey) {
      throw new Error('private_key is required to sign the transaction');
    }

    const account = this.web3.eth.accounts.privateKeyToAccount(privateKey);
    const chainId = await this.web3.eth.getChainId();
    const latestBlock = await this.web3.eth.getBlock('latest');
    
    // Check if EIP-1559 is supported
    const eip1559Supported = latestBlock && 'baseFeePerGas' in latestBlock;
    
    const nonce = await this.web3.eth.getTransactionCount(account.address);
    
    const common = {
      from: account.address,
      nonce: nonce,
      gas: 1000000,
      chainId: chainId,
    };

    let tx: any;

    if (eip1559Supported) {
      try {
        // Use eth_feeHistory to estimate fees
        const feeHistory = await this.web3.eth.getFeeHistory(10, 'latest', [50]);
        const rewards = feeHistory.reward?.map((r: any) => r[0]).filter((r: any) => r) || [];
        
        let tip = rewards.length > 0 
          ? BigInt(rewards.sort((a: any, b: any) => Number(a) - Number(b))[Math.floor(rewards.length / 2)])
          : BigInt(this.web3.utils.toWei('2', 'gwei'));
        
        const baseFees = feeHistory.baseFeePerGas || [];
        const nextBase = baseFees.length > 0 
          ? BigInt(baseFees[baseFees.length - 1])
          : BigInt(latestBlock.baseFeePerGas || 0);
        
        // Enforce minimum tip
        if (tip < BigInt(this.web3.utils.toWei('1', 'gwei'))) {
          tip = BigInt(this.web3.utils.toWei('1', 'gwei'));
        }
        
        // Robust cap: 2x next base fee + median tip
        const maxFee = nextBase * 2n + tip;
        
        tx = {
          ...common,
          maxPriorityFeePerGas: tip.toString(),
          maxFeePerGas: maxFee.toString(),
          type: 2,
          data: this.contract.methods.register(agentId, metadata).encodeABI(),
          to: this.contract.options.address,
        };
      } catch (error) {
        // Fallback to simple heuristic
        const baseFee = BigInt(latestBlock.baseFeePerGas || 0);
        const tip = BigInt(this.web3.utils.toWei('2', 'gwei'));
        const maxFee = baseFee * 2n + tip;
        
        tx = {
          ...common,
          maxPriorityFeePerGas: tip.toString(),
          maxFeePerGas: maxFee.toString(),
          type: 2,
          data: this.contract.methods.register(agentId, metadata).encodeABI(),
          to: this.contract.options.address,
        };
      }
    } else {
      const gasPrice = await this.web3.eth.getGasPrice();
      tx = {
        ...common,
        gasPrice: gasPrice.toString(),
        data: this.contract.methods.register(agentId, metadata).encodeABI(),
        to: this.contract.options.address,
      };
    }

    const signed = await account.signTransaction(tx);
    const receipt = await this.web3.eth.sendSignedTransaction(signed.rawTransaction as string);
    
    return (receipt as any).transactionHash;
  }

  async getEntry(address: string): Promise<any> {
    return await this.contract.methods.getEntry(address).call();
  }

  async getAllEntries(fromBlock: number = 0, toBlock: string | number = 'latest'): Promise<ContractEntry[]> {
    // Scan Registered events to discover all owners
    const topic0 = this.web3.utils.keccak256('Registered(address,string,string,uint256)');
    
    const logs = await this.web3.eth.getPastLogs({
      address: this.contract.options.address,
      fromBlock: fromBlock,
      toBlock: toBlock,
      topics: [topic0],
    });

    const ownersSet = new Set<string>();
    const owners: string[] = [];

    for (const log of logs) {
      // owner is the first indexed topic (topics[1])
      if (typeof log !== 'string' && log.topics && log.topics.length > 1) {
        const t1 = log.topics[1];
        const owner = this.web3.utils.toChecksumAddress('0x' + t1.slice(-40));
        const ownerLower = owner.toLowerCase();
        
        if (!ownersSet.has(ownerLower)) {
          ownersSet.add(ownerLower);
          owners.push(owner);
        }
      }
    }

    const entries: ContractEntry[] = [];
    
    for (const owner of owners) {
      try {
        const entry = await this.contract.methods.getEntry(owner).call();
        entries.push({
          owner: entry[0],
          agentId: entry[1],
          metadata: entry[2],
          updatedAt: entry[3],
        });
      } catch (error) {
        // Skip failed entries
        continue;
      }
    }

    return entries;
  }
}